### PR TITLE
Common code to support read and write RPC.

### PR DESCRIPTION
**Pull request by smjn**
* Code to create ShareSnapshot type records.
* Update PersisterStateManager main handler.
* Add lifecycle methods in ShareCoordinator.
* Create timelinehashmaps in ShareCoordinatorShard to store soft state.

### CODE REVIEW

**Comment by chirag-wadhwa5** [comment]
LGTM

**Review by chirag-wadhwa5 — APPROVED** [review]

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/share/ShareCoordinatorShard.java [hunk 1]
```diff
@@ -22,23 +22,37 @@
...
+import java.util.Map;
+
 public class ShareCoordinatorShard implements CoordinatorShard<Record> {
   private final Logger log;
   private final Time time;
   private final CoordinatorTimer<Void, Record> timer;
   private final ShareCoordinatorConfig config;
   private final CoordinatorMetrics coordinatorMetrics;
   private final CoordinatorMetricsShard metricsShard;
+  // coord key -> ShareShap
```
Comment (AndrewJSchofield) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/share/ShareCoordinatorShard.java:53]: ShareShap looks like a typo

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/share/ShareCoordinatorShard.java [hunk 2]
```diff
@@ -153,7 +171,34 @@ public void onUnloaded() {
 
   @Override
   public void replay(long offset, long producerId, short producerEpoch, Record record) throws RuntimeException {
+    ApiMessageAndVersion key = record.key();
+    ApiMessageAndVersion value = record.value();
+
+    switch (key.version()) {
+      case 0: // ShareSnapshot
```
Comment (AndrewJSchofield) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/share/ShareCoordinatorShard.java:178]: It's always struck me as a little strange that there are no constant definitions for these values. I think that would be a useful idea to introduce with the share coordinator. Possibly in the ShareCoordinator class.

**Review by AndrewJSchofield — APPROVED** [review]
A couple of tiny comments, but looks good to me.

**Review by AndrewJSchofield — APPROVED** [review]

**Pull request merged**